### ISSUE TITLE
Curious case of tmux 2.6 vs. htop-git

### ISSUE THREAD

**Issue by mkoskar** [tmux/tmux#1112]
Versions:
```
Linux mirci 4.13.4-1-ARCH #1 SMP PREEMPT Thu Sep 28 08:39:52 CEST 2017 x86_64 GNU/Linux
tmux 2.6
htop - master git (e3f65c8ec2fe81c50514dcf21c335e5d39c1a223)
```

I have this test script:
```bash
#!/usr/bin/env bash

socket='test'
session='test'

tmux -vv -L "$socket" -f /dev/null new -d -s "$session" \
    \; setw remain-on-exit on \
    \; splitw -h htop

exec tmux -L "$socket" attach -t "=$session"
```

Running above you'll find that tmux server is killed with following as last line in `tmux-server-*.log`:
```
1507732247.504665 fatal: ioctl failed: Bad file descriptor
```

1) downgrading `htop` to latest stable release of `2.0.2` resolves the issue
2) downgrading `tmux` to `2.5` also resolves the issue
3) if I add `sleep 1` just before `attach` server isn't killed but `htop` is dead
4) if I remove `-d` from `new-session` command, i.e., let it start on foreground all seems to be well
5) if I start tmux server with `strace -D -ff -o strace ...`, server doesn't crash itself but I get strace output for htop:

```
write(1, "\33[34h\33[?25h", 11)         = 11
write(1, "\33[?1000l", 8)               = 8
write(1, "\33[39;49m\33[m\17\33[24;1H\33[?1049l\r\33[?1"..., 35) = 35
ioctl(1, SNDCTL_TMR_STOP or TCSETSW, {B38400 opost isig icanon echo ...}) = 0
write(2, "htop: ", 6)                   = 6
write(2, "Success\n", 8)                = 8
exit_group(1)                           = ?
+++ exited with 1 +++
```

Unfortunately I haven't had more time to investigate any further, but at least one other person on IRC could reproduce the issue with above test script.

**Comment by nicm** [comment]
What is the actual problem?



On 11 Oct 2017 3:23 pm, "Miroslav Koškár" <notifications@github.com> wrote:

Versions:

Linux mirci 4.13.4-1-ARCH #1 SMP PREEMPT Thu Sep 28 08:39:52 CEST 2017
x86_64 GNU/Linux
tmux 2.6
htop - master git (e3f65c8ec2fe81c50514dcf21c335e5d39c1a223)

I have this test script:

#!/usr/bin/env bash

socket='test'
session='test'

tmux -vv -L "$socket" -f /dev/null new -d -s "$session" \
    \; setw remain-on-exit on \
    \; splitw -h htop
exec tmux -L "$socket" attach -t "=$session"


   1. downgrading htop to latest stable release of 2.0.2 resolves the issue
   2. downgrading tmux to 2.5 also resolves the issue
   3. if I add sleep 1 just before attach server isn't killed but htop is
   dead
   4. if I remove -d from new-session command, i.e., let it start of
   foreground all seems to be well
   5. if I start tmux server with strace -D -ff -o strace ..., server
   doesn't crash itself but I get strace output for htop:

write(1, "\33[34h\33[?25h", 11)         = 11
write(1, "\33[?1000l", 8)               = 8
write(1, "\33[39;49m\33[m\17\33[24;1H\33[?1049l\r\33[?1"..., 35) = 35
ioctl(1, SNDCTL_TMR_STOP or TCSETSW, {B38400 opost isig icanon echo ...}) = 0
write(2, "htop: ", 6)                   = 6
write(2, "Success\n", 8)                = 8
exit_group(1)                           = ?
+++ exited with 1 +++

Unfortunately I haven't had more time to investigate any further, but at
least one person on IRC could reproduce the issue with about test script.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/tmux/tmux/issues/1112>, or mute the thread

**Comment by mkoskar** [comment]
@nicm sorry, I've messed up and left out (now edited and added):

Running above you'll find that tmux server is killed with following as last line in `tmux-server-*.log`:
```
1507732247.504665 fatal: ioctl failed: Bad file descriptor
```

**Comment by nicm** [comment]
Show me the entire log please.

**Comment by mkoskar** [comment]
[tmux-client-16537.log](https://github.com/tmux/tmux/files/1375982/tmux-client-16537.log)
[tmux-out-16539.log](https://github.com/tmux/tmux/files/1375983/tmux-out-16539.log)
[tmux-server-16539.log](https://github.com/tmux/tmux/files/1375984/tmux-server-16539.log)

**Comment by mkoskar** [comment]
Just to add, I've tried `tmux git master` with same result.
I'm using `urxvt` terminal with `TERM=rxvt-unicode-256color`, tried also from `xterm` with same result.

**Comment by nicm** [comment]
OK I suspect that htop is exiting, probably when tmux tries to resize it (2.5 would not resize when you attach, but 2.6 will), which closes the slave side of the pty which on your platform makes ioctl() return EBADF on the master side when tmux tries TIOCSWINSZ. ISTR the same issue before but either it was in a different place or the person who reported it disappeared so it was never fixed.

Anyway, try this please: [x.diff.txt](https://github.com/tmux/tmux/files/1376406/x.diff.txt)

**Comment by mkoskar** [comment]
I see, ok but the patch doesn't apply cleanly to master, nor does it to 2.6 tag. Hmm

**Comment by nicm** [comment]
Actually, this looks like a bad merge - the wp->fd != -1 check is missing in GitHub. Try master now, d10def5b0b5a4d000b1a34efd32cc4a2613f66c7 might fix it.

**Comment by mkoskar** [comment]
Yep, git master fixes it (for the record I've tried fb02df66).
So now tmux doesn't crash itself, yet htop is dead.
Your theory with it crashing on resize (`SIGWINCH`) seems plausible, so I'll raise it there.

Thanks!

**Comment by nicm** [comment]
Great thanks.
